### ISSUE TITLE
Token invalid for a minute... then becomes valid?

### ISSUE THREAD

**Issue by karam94** [tymondesigns/jwt-auth#1337]
Building a project as an API with Laravel, that utilises JWT-Auth.

On my dev machine, I never encountered this issue.
However after deploying to Heroku, this issue has become apparent.

Basically, I log in and my client receives a JWT token.
However, it takes a good 1-2 minutes before that JWT token is actually usable.

For example, I can keep sending the token I receive across using postman and I'll keep receiving "token_invalid" for a minute or two, then eventually magically I'll actually receive the user data!

What on earth is causing this? It's driving me crazy and is holding back the actual deployment of the application.

Thanks!

**Comment by karam94** [comment]
Solved the problem.
Debugged the JWT being returned from Heroku and the 'iat' claim and others were all set in the future.
Obviously, my Heroku server was on the correct unix time otherwise it wouldn't return invalid token as the token was in the future.
So, I refreshed my Macbook Pro's time and it was brought back about 30 seconds and the problem no longer persists...

However, I do worry if any potential users have a slightly off computer clock, this could result in them trying to log in and requesting a JWT token in the future/past. I don't think this is a security risk as such, but I would like to add some kind of leeway on the time. For example, set the claims of the token returned to 5 minutes or so in the past.

Is this possible? Is this a bad idea? If so, why? And how do I prevent users suffering from this discrepancy?

**Comment by jampack** [comment]
@karam94 i dont think it has anything to do with clients time or anything unless explicitly done.
'iat' is the time on server the token is issued at.
'exp' is the time on server that the token will expire on.
'nbf' is the time on server that the token cant be used before.
by package defaults 'iat' & 'nbf' are exactly the same and exactly the current time on server so that the token can be used right after issuance.
When you say the token  is not usable for a while and then becomes valid, that has all to do with 'nbf' => 'not before' coz messing with 'iat' to make it a future time doesn't make any sense, but the bottom line is client time has nothing to do with tokens time by package defaults

**Comment by jampack** [comment]
@karam94 and for the security risk if the time had to be calculated based on clients machine then they can always alter their time to make use of a token that has expired long ago according to 'exp'
